tapdb/sqlc/queries/universe.sql

```diff
@@ -370,4 +370,87 @@ ON CONFLICT(namespace)
 -- name: QueryFederationUniSyncConfigs :many
 SELECT namespace, asset_id, group_key, proof_type, allow_sync_insert, allow_sync_export
 FROM federation_uni_sync_config
-ORDER BY group_key NULLS LAST, asset_id NULLS LAST, proof_type;
\ No newline at end of file
+ORDER BY group_key NULLS LAST, asset_id NULLS LAST, proof_type;
+
+-- name: UpsertFederationProofSyncLog :one
+INSERT INTO federation_proof_sync_log as log (
+    status, timestamp, sync_direction, proof_leaf_id, universe_root_id,
+    servers_id
+) VALUES (
+    @status, @timestamp, @sync_direction,
+    (
+        -- Select the leaf id from the universe_leaves table.
+        SELECT id
+        FROM universe_leaves
+        WHERE leaf_node_namespace = @leaf_namespace
+            AND minting_point = @leaf_minting_point_bytes
+            AND script_key_bytes = @leaf_script_key_bytes
+        LIMIT 1
+    ),
+    (
+        -- Select the universe root id from the universe_roots table.
+        SELECT id
+        FROM universe_roots
+        WHERE namespace_root = @universe_id_namespace
+        LIMIT 1
+    ),
+    (
+        -- Select the server id from the universe_servers table.
+        SELECT id
+        FROM universe_servers
+        WHERE server_host = @server_host
+        LIMIT 1
+    )
+) ON CONFLICT (sync_direction, proof_leaf_id, universe_root_id, servers_id)
+DO UPDATE SET
+    status = EXCLUDED.status,
+    timestamp = EXCLUDED.timestamp,
+    -- Increment the attempt counter.
+    attempt_counter = CASE
+       WHEN @bump_sync_attempt_counter = true THEN log.attempt_counter + 1
+       ELSE log.attempt_counter
+    END
+RETURNING id;
+
+-- name: QueryFederationProofSyncLog :many
```

Comment:
We should add some basic unit tests for the upsert and query to make sure they work as expected, since they look a bit more involved (especially with the `ON CONFLICT ... DO UPDATE`, that always tends to have surprising side effects, at least in my experience).

Reply:
Unit tests added :+1: